primer3/thermoanalysis.pyx

```diff
@@ -949,3 +958,1147 @@ cdef class ThermoAnalysis:
+            ):
+                output_dict[f'PRIMER_PAIR_{i}_PRODUCT_TM'] = \
+                    retval[0].best_pairs.pairs[i].product_tm
+                output_dict[f'PRIMER_PAIR_{i}_PRODUCT_TM_OLIGO_TM_DIFF'] = \
+                    retval[0].best_pairs.pairs[i].product_tm_oligo_tm_diff
+                output_dict[f'PRIMER_PAIR_{i}_T_OPT_A'] = retval[0].best_pairs.pairs[i].t_opt_a
+            else:
+                output_dict[f'PRIMER_PAIR_{i}_PRODUCT_TM'] = \
+                retval[0].best_pairs.pairs[i].product_tm
+
+            # Print the primer pair template mispriming
+            if (
+                (global_settings_data[0].thermodynamic_template_alignment == 0) and
+                (retval[0].best_pairs.pairs[i].template_mispriming != ALIGN_SCORE_UNDEF)
+            ):
+                output_dict[f'PRIMER_PAIR_{i}_TEMPLATE_MISPRIMING'] = \
+                    retval[0].best_pairs.pairs[i].template_mispriming
+            # Print the primer pair template mispriming. Thermodynamic approach.
+            if (
+                    (global_settings_data[0].thermodynamic_template_alignment == 1) and
+                    (retval[0].best_pairs.pairs[i].template_mispriming != ALIGN_SCORE_UNDEF)
+            ):
+                output_dict[f'PRIMER_PAIR_{i}_TEMPLATE_MISPRIMING_TH'] = \
+                    retval[0].best_pairs.pairs[i].template_mispriming
+
+            # Print primer secondary structures*/
+            if (
+                 (global_settings_data[0].show_secondary_structure_alignment == 1) and
+                (retval[0].best_pairs.pairs[i].template_mispriming_struct != NULL)
+            ):
+                sqtemp_b = <bytes> retval[0].best_pairs.pairs[i].template_mispriming_struct
+                output_dict[f'PRIMER_PAIR_{i}_TEMPLATE_MISPRIMING_STUCT'] = \
+                    sqtemp_b.decode('utf8')
+        # End of print parameters of primer pairs
+    # End of the for big loop printing all data
+    return output_dict
+
+
+def _p3_global_design_cleanup():
+    # Free any remaining global Primer3 objects
+    global global_settings_data
+    global sequence_args_data
+
+    destroy_thal_structures()
+    if global_settings_data != NULL:
+        # Free memory for previous global settings
+        p3_destroy_global_settings(global_settings_data)
+        global_settings_data = NULL
+
+    if sequence_args_data != NULL:
+        # Free memory for previous seq args
+        destroy_seq_args(sequence_args_data)
+        sequence_args_data = NULL
+
+atexit.register(_p3_global_design_cleanup)
+
+
+class Singleton(type):
+    _instances = {}  # type: ignore
+
+    def __call__(cls, *args, **kwargs):
+        if cls not in cls._instances:
+            cls._instances[cls] = super(Singleton, cls).__call__(
+                *args,
+                **kwargs,
+            )
+        return cls._instances[cls]
+
+
+class ThermoAnalysis(_ThermoAnalysis, metaclass=Singleton):
+
+    def calcHeterodimer(
```

Comment:
I think in theory you could just add these shims in `__init__` like:

```
def __init__(self):
    self.calcHeterodimer = self.calc_heterodimer
```
I believe that should preserve the function signatures, docstring, etc through reference.

Also, any reason not to do this on the `_ThermoAnalysis` base class?

Reply:
I chose to have the base `_ThermoAnalysis` class to not have the deprecation warnings.
I also didn't do the `self.calcHeterodimer = self.calc_heterodimer` because our deprecation warning goal.

By leaving them on the subclass, removing the deprecation warnings is just a matter of deleting all of the methods on `ThermoAnalysis`.   It's a 50/50 call to say where they belong.

Reply:
Ah ok, current solution makes sense then